# Advanced Dataframes

In this lesson we will continue working with pandas DataFrames, and explore some more complex DataFrame manipulations.

In [1]:
import pandas as pd
import numpy as np

np.random.seed(123)

In [2]:
# Create list of values for names column.

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# Randomly generate arrays of scores for each student for each subject.
# Note that all the values need to have the same length here.

math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

In [3]:
# Construct the DataFrame using the above lists and arrays.

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades,
                   'classroom': np.random.choice(['A', 'B'], len(students))})

In [4]:
df.head()

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A
3  Billy    98       96       88         B
4    Ada    77       92       98         A

## Creating Dataframes

### From Lists and Dictionaries

There are several ways to create dataframes, we've already seen how we can create a dataframe from a dictionary:

In [5]:
pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})

A  B
0  1  4
1  2  5
2  3  6

The keys in the passed dictionary will be the column names, and the values are the data points that make up each column.

We can also create dataframes from a 2d data structure, either a numpy array or a list of lists:

In [6]:
pd.DataFrame([[1, 2, 3], [4, 5, 6 ]])

0  1  2
0  1  2  3
1  4  5  6

In [7]:
array = np.array([[1, 2, 3], [4, 5, 6]])

pd.DataFrame(array, columns=['this', 'that', 'the other'])

this  that  the other
0     1     2          3
1     4     5          6

Notice here that we had to specify the names of the columns ourselves.

### From PyDataset.

In some of the exercises, you'll need to load several datasets using the `pydataset` library. (If you get an error when trying to run the import below, use `pip` to install the `pydataset` package.) When the instructions say to load a dataset from PyDataset, you will need to do the following:

The following import is necessary to access PyDataset datasets:
```python
from pydataset import data
```

Running this code snippet will show you the valuable information doc on the dataset:
```python
data(df_string_name, show_doc=True)
```

Running this code snippet will load the dataset for use as a pandas DataFrame:
```python
df = data(df_string_name)
```

There are 757 available datasets using pydataset. Running the following code snippet in a cell will return a DataFrame with all of your options:
```python
data()
```

In [8]:
# Load the dataset and store it in the variable mpg.

from pydataset import data
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

You can view the documentation for the dataset, and any pydata dataset, by setting show_doc to True. This outputs valuable context for your dataset.

```python
data('mpg', show_doc=True) 
```

### From SQL

We can use the `read_sql` method to create a dataframe based on the results of a SQL query. To do this, we need to tell pandas how to connect to the database we are querying. The way we communicate this to pandas is with a specially formatted *connection string*.

In addition, whenever we want to connect to a database from our python code (other programming languages are similar), we will need a **driver**, a bit of software that handles the details of the database connection.

In order to connect to mysql, we'll install the `pymysql` driver packages:

`python -m pip install pymysql`

Once those are installed, we can create the connection string. In general, database connection urls will have this format:

```python
protocol://[user[:password]@]hostname/[database_name]
```

Here's an example of what one would look like:

```python
mysql+pymysql://codeup:p@assw0rd@123.123.123.123/some_db
```

Another thing we need to consider is that we don't want to publish our database credentials to github, however, we will need access to these values in our code in order to create the connection string defined above.

In order to accomplish this, we can define several variables in a file named `env.py` that contain the sensitive data, add `env.py` to our `.gitignore` file, and then import those values into another script. 

**Be 100% sure to add `env.py` to this specific repository's `.gitignore` file, even and especially, if you have already added `env.py` to your global .gitignore file. This will protect the env file for people who clone this project (like collaborators)**

In [9]:
# pd.read_sql(query, url)

## Method 1 
    -see method 2 in exercise I
    -I commented out the below cells so only one method is run in the notebook

from env import host, user, password

url = f'mysql+pymysql://{user}:{password}@{host}/employees'

Once this url is defined, we can use it with the `read_sql` function to have pandas treat the results of a SQL query as a dataframe.

pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

It is common to have longer SQL queries that we want to read into python, and an example of how we might break a query into several lines is below:

sql = '''
SELECT
    emp_no,
    first_name,
    last_name
FROM employees
WHERE gender = 'F'
LIMIT 100
'''

employees = pd.read_sql(sql, url)
employees.head()

!!!danger "Passwords and Sensitive Information"
    Don't add and commit files with passwords or other sensitive information in them to a git repository!

query = '''
SELECT
    t.title as title,
    d.dept_name as dept_name
FROM titles t
JOIN dept_emp USING (emp_no)
JOIN departments d USING (dept_no)
LIMIT 100
'''

title_dept = pd.read_sql(query, url)
title_dept.head()

## Exercises I

Run `python -m pip install pymysql` from your terminal to install pymysql.

Create a notebook or python script named `advanced_dataframes` to do your work in for these exercises.

1. Run `python -m pip install pymysql` from your terminal to install the mysql client (any folder is fine)
1. cd into your exercises folder for this module and run `echo env.py >> .gitignore`
1. Create a function named `get_db_url`. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

2. Use your function to obtain a connection to the `employees` database.

3. Once you have successfully run a query:

    a. Intentionally make a typo in the database url. What kind of error message do you see?
    
    b. Intentionally make an error in your SQL query. What does the error message look like?

4. Read the `employees` and `titles` tables into two separate DataFrames.

5. How many rows and columns do you have in each DataFrame? Is that what you expected?

6. Display the summary statistics for each DataFrame.

7. How many unique titles are in the `titles` DataFrame?

8. What is the oldest date in the `to_date` column? 

9. What is the most recent date in the `to_date` column?

In [10]:
# class
# from env improt host, username, password


# def get_db_url(username, host, password, db):
#      return f'mysql+pymysql://{user}:{password}@{host}/{db}'


# url = get_db_url(username, host, password, 'employees')

# pd.read_sql()

----------------------

In [11]:
# This function is in the env.py file
# def get_db_url(db):
#     return f'mysql+pymysql://{user}:{password}@{host}/{db}'

In [2]:
import env

In [13]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', env.get_db_url('employees') )



emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

In [14]:
#m'ed out example from above. 
#Note the replacement of 'url' with 'env.get_db_url('employees')'
sql = '''
SELECT
    emp_no,
    first_name,
    last_name
FROM employees
WHERE gender = 'F'
LIMIT 100
'''

employees = pd.read_sql(sql, env.get_db_url('employees'))
employees.head()

emp_no first_name  last_name
0   10002    Bezalel     Simmel
1   10006     Anneke    Preusig
2   10007    Tzvetan  Zielinski
3   10009     Sumant       Peac
4   10010  Duangkaew   Piveteau

In [15]:
#2nd m'ed out example from pre-exercise I
query = '''
SELECT
    t.title as title,
    d.dept_name as dept_name
FROM titles t
JOIN dept_emp USING (emp_no)
JOIN departments d USING (dept_no)
LIMIT 100
'''

title_dept = pd.read_sql(query, env.get_db_url('employees'))
title_dept.head()

title         dept_name
0         Staff  Customer Service
1  Senior Staff  Customer Service
2         Staff  Customer Service
3  Senior Staff  Customer Service
4         Staff  Customer Service

#5a Intentionally make a typo in the database ur. What kind of error message do you see?
sql = '''
SELECT
    emp_no,
    first_name,
    last_name
FROM employees
WHERE gender = 'F'
LIMIT 100
'''

employees = pd.read_sql(sql, env.get_db_url('employeees'))
employees.head()

#OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'kalpana_1820'@'%' to database 'employeees'")
#(Background on this error at: https://sqlalche.me/e/14/e3q8)

#5b Intentionally make a typo in the database ur. What kind of error message do you see?
sql = '''
SELECT
    emp_no,
    first_name,
    last_name,
FROM employees
WHERE gender = 'F'
LIMIT 100
'''

employees = pd.read_sql(sql, env.get_db_url('employees'))
employees.head()

#removed , after emp_no, just returned table without emp_no
#added , after last_name and it returned:

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM employees\nWHERE gender = 'F'\nLIMIT 100' at line 5")
[SQL: 
SELECT
    emp_no,
    first_name,
    last_name,
FROM employees
WHERE gender = 'F'
LIMIT 100
]
(Background on this error at: https://sqlalche.me/e/14/f405)

In [16]:
#it appears if it cannot connect, it is an Operational Error
#if the code makes no sense inside the connected database, it is a ProgrammingError

In [17]:
query = '''
SELECT
    t.title as title,
    e.first_name,
    e.last_name
FROM employees e
JOIN titles t USING (emp_no)
LIMIT 100
'''

titles = pd.read_sql(query, env.get_db_url('employees'))
titles.head()

title first_name last_name
0  Senior Engineer     Georgi   Facello
1            Staff    Bezalel    Simmel
2  Senior Engineer      Parto   Bamford
3         Engineer  Chirstian   Koblick
4  Senior Engineer  Chirstian   Koblick

In [18]:
query = '''
SELECT
    *
FROM employees
LIMIT 100
'''

employees = pd.read_sql(query, env.get_db_url('employees'))
employees.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [19]:
query = '''
SELECT
    *
FROM titles
'''

titles = pd.read_sql(query, env.get_db_url('employees'))
titles.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

In [20]:
query = '''
SELECT
    *
FROM employees
JOIN titles USING (emp_no)
LIMIT 100
'''

employees_and_titles = pd.read_sql(query, env.get_db_url('employees'))
employees_and_titles.head()

emp_no  birth_date first_name last_name gender   hire_date  \
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   

             title   from_date     to_date  
0  Senior Engineer  1986-06-26  9999-01-01  
1            Staff  1996-08-03  9999-01-01  
2  Senior Engineer  1995-12-03  9999-01-01  
3         Engineer  1986-12-01  1995-12-01  
4  Senior Engineer  1995-12-01  9999-01-01

In [21]:
employees.shape #yes, cuz I limited 100

(100, 6)

In [22]:
query = '''
SELECT
    *
FROM employees
'''

employees = pd.read_sql(query, env.get_db_url('employees'))
employees.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [23]:
employees.shape #without limiting to 100, I get 300024, 6

(300024, 6)

In [24]:
query = '''
SELECT
    *
FROM employees
LIMIT 100
'''

employees = pd.read_sql(query, env.get_db_url('employees'))
employees.head()
#limiting to 100 again for sake of time

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [25]:
employees.describe(), employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   emp_no      100 non-null    int64 
 1   birth_date  100 non-null    object
 2   first_name  100 non-null    object
 3   last_name   100 non-null    object
 4   gender      100 non-null    object
 5   hire_date   100 non-null    object
dtypes: int64(1), object(5)
memory usage: 4.8+ KB


(             emp_no
 count    100.000000
 mean   10050.500000
 std       29.011492
 min    10001.000000
 25%    10025.750000
 50%    10050.500000
 75%    10075.250000
 max    10100.000000,
 None)

In [26]:
employees.dtypes

emp_no         int64
birth_date    object
first_name    object
last_name     object
gender        object
hire_date     object
dtype: object

In [27]:
titles.info(), titles.describe(), titles.dtypes, titles.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


(None,
               emp_no
 count  443308.000000
 mean   253075.034430
 std    161853.292613
 min     10001.000000
 25%     84855.750000
 50%    249847.500000
 75%    424891.250000
 max    499999.000000,
 emp_no        int64
 title        object
 from_date    object
 to_date      object
 dtype: object,
 (443308, 4))

In [28]:
titles.title.value_counts()

Engineer              115003
Staff                 107391
Senior Engineer        97750
Senior Staff           92853
Technique Leader       15159
Assistant Engineer     15128
Manager                   24
Name: title, dtype: int64

In [29]:
titles.title.unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object)

In [30]:
len(titles.title.unique())

7

In [31]:
np.unique(titles.title)

array(['Assistant Engineer', 'Engineer', 'Manager', 'Senior Engineer',
       'Senior Staff', 'Staff', 'Technique Leader'], dtype=object)

In [32]:
len(np.unique(titles.title))

7

In [33]:
#pd.to_datetime(titles.to_date, yearfirst=False) 
#OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 9999-01-01 00:00:00

In [34]:
pd.to_datetime(titles.to_date, errors='coerce')


0               NaT
1               NaT
2               NaT
3        1995-12-01
4               NaT
            ...    
443303   1992-08-29
443304          NaT
443305          NaT
443306   1998-12-27
443307          NaT
Name: to_date, Length: 443308, dtype: datetime64[ns]

In [35]:
titles.dtypes

emp_no        int64
title        object
from_date    object
to_date      object
dtype: object

In [36]:
#titles['to_date'].nsmallest() nsmallest/nlargest have to be int, use min max

In [37]:
titles.to_date.max()

datetime.date(9999, 1, 1)

In [38]:
titles.to_date.min()

datetime.date(1985, 3, 1)

In [39]:
#titles.to_date.idxmin() #doesn't work with non-intergers

In [40]:
import datetime #class

In [41]:
past_employees = titles[titles.to_date != datetime.date(9999, 1, 1)] #class

In [42]:
past_employees.max() #class, missing one line don't worry about it

emp_no                 499998
title        Technique Leader
from_date          2002-06-30
to_date            2002-08-01
dtype: object

## Indexing and Subsetting

Like the pandas Series object, the pandas DataFrame object supports both position- and label-based indexing using the indexing operator `[]`.

I will demonstrate concrete examples of indexing using the indexing operator `[]` alone and with the `.loc` and `.iloc` attributes below.

### `[]`

I can pass a list of columns from a DataFrame to the indexing operator (aka bracket notation) to return a subset of my original DataFrame.

In [43]:
df[1:5]

name  math  english  reading classroom
1   Jane    88       79       67         B
2  Suzie    94       74       95         A
3  Billy    98       96       88         B
4    Ada    77       92       98         A

In [44]:
df[df.math > 85]

name  math  english  reading classroom
1     Jane    88       79       67         B
2    Suzie    94       74       95         A
3    Billy    98       96       88         B
7    Marie    93       63       90         A
8   Albert    92       62       87         A
10   Isaac    92       99       93         B
11    Alan    92       62       72         A

In [45]:
# Choose only two columns for my subset.

df[['name', 'classroom']].head()

name classroom
0  Sally         A
1   Jane         B
2  Suzie         A
3  Billy         B
4    Ada         A

In [46]:
# I can pass a boolean Series to the indexing operator as a selector.

bools = df.name.str.startswith('A')
bools

0     False
1     False
2     False
3     False
4      True
5     False
6     False
7     False
8      True
9     False
10    False
11     True
Name: name, dtype: bool

In [47]:
df[bools]

name  math  english  reading classroom
4      Ada    77       92       98         A
8   Albert    92       62       87         A
11    Alan    92       62       72         A

### `.loc`

We can use the `.loc` attribute to select specific rows AND columns by index label. The index label can be a number, but it can also be a string label. This method offers a lot of flexibility! **The .loc attribute's indexing is inclusive and uses an index label, not position.**

```python
df.loc[row_indexer, column_indexer]
```

In [48]:
df.head()

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A
3  Billy    98       96       88         B
4    Ada    77       92       98         A

In [49]:
# Select all the rows and a subset of columns; notice the inclusive behavior of the indexing.

df.loc[:, 'math':'reading']

math  english  reading
0     62       85       80
1     88       79       67
2     94       74       95
3     98       96       88
4     77       92       98
5     79       76       93
6     82       64       81
7     93       63       90
8     92       62       87
9     69       80       94
10    92       99       93
11    92       62       72

In [50]:
df.loc[0:4, 'math':'reading']

math  english  reading
0    62       85       80
1    88       79       67
2    94       74       95
3    98       96       88
4    77       92       98

### `.iloc`

We can use the `.iloc` attribute to select specific rows and colums by index position. `.iloc` does not accept a boolean Series as a selector like `.loc` does. **It takes in integers representing index position and is NOT inclusive.**

```python
df.iloc[row_indexer, column_indexer]
```

We can select rows by integer position:

In [51]:
# Notice the exclusive behavior of the indexing.

df.iloc[:3]

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A

We can also specify which columns we want to select:

In [52]:
df.iloc[:3, 1:3]

math  english
0    62       85
1    88       79
2    94       74

Here we select the first 3 rows (everything up to but not including the index of 3), and the second and third columns (starting from the index of 1 up to but not including the index of 3).

## Aggregating

### `.agg`

The `.agg` method lets us specify a way to aggregate a series of numerical values. We pass an aggregate function or list of functions to the method that we want applied to a Series.

In [53]:
df.math.median()

90.0

In [54]:
df.math.min()

62

In [55]:
df.reading.agg('min')

67

While on the surface this seems pretty simple, `.agg` is capable of providing more detailed aggregations:

In [56]:
df.head()

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A
3  Billy    98       96       88         B
4    Ada    77       92       98         A

In [57]:
df[['english', 'reading', 'math']].mean() # outer [] for pytong inner [] list

english    77.666667
reading    86.500000
math       84.833333
dtype: float64

In [58]:
df[['english', 'reading', 'math']].agg(['mean', 'min', 'max'])

english  reading       math
mean  77.666667     86.5  84.833333
min   62.000000     67.0  62.000000
max   99.000000     98.0  98.000000

### `.groupby`

The `.groupby` method is used to create a grouped object, which we can then apply an aggregation on. For example, if we wanted to know the highest math grade from each classroom:

In [59]:
df.groupby('classroom') #k, now waht do you want?

In [60]:
df[["classroom", "math", "reading", "english"]].groupby("classroom").max()

math  reading  english
classroom                        
A            94       98       92
B            98       93       99

In [61]:
df.groupby('classroom').math.max()

classroom
A    94
B    98
Name: math, dtype: int64

We can use `.agg` here to, to see multiple aggregations:

In [62]:
df.groupby('classroom')[['math', 'reading']].agg(['min', 'mean', 'max'])

math             reading            
           min    mean max     min    mean max
classroom                                     
A           62  82.625  94      72  87.125  98
B           79  89.250  98      67  85.250  93

In [63]:
df[['math']] #pretty dataframe

math
0     62
1     88
2     94
3     98
4     77
5     79
6     82
7     93
8     92
9     69
10    92
11    92

In [64]:
df.math #less pretty series

0     62
1     88
2     94
3     98
4     77
5     79
6     82
7     93
8     92
9     69
10    92
11    92
Name: math, dtype: int64

In [65]:
df.head()

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A
3  Billy    98       96       88         B
4    Ada    77       92       98         A

We can group by multiple columns as well. To demonstrate, we'll create a boolean column named `passing_math`, then group by the combination of our new feature, `passing_math`, and the classroom and calculate the average reading grade and the number of individuals in each subgroup. 

Let's break this problem down and code it step-by-step.

#### `np.where`

First, we can create the new `passing_math` column using a handy NumPy function called `np.where`. It will allow us to base the new column values on whether the values in an existing column, `math`, meet a condition.

```python
np.where(condition, this_where_True, this_where_False)
```

In [66]:
# Create the new column based on an existing column.

df['passing_math'] = np.where(df.math < 70, 'failing', 'passing')
df

name  math  english  reading classroom passing_math
0     Sally    62       85       80         A      failing
1      Jane    88       79       67         B      passing
2     Suzie    94       74       95         A      passing
3     Billy    98       96       88         B      passing
4       Ada    77       92       98         A      passing
5      John    79       76       93         B      passing
6    Thomas    82       64       81         A      passing
7     Marie    93       63       90         A      passing
8    Albert    92       62       87         A      passing
9   Richard    69       80       94         A      failing
10    Isaac    92       99       93         B      passing
11     Alan    92       62       72         A      passing

Next, we will group by the `passing_math` and `classroom` columns and use the `.agg` method to calculate the average reading grade and the number of students.

In [67]:
grade_groups = df.groupby(['passing_math', 'classroom']).reading.agg(['mean', 'count'])
grade_groups

mean  count
passing_math classroom                  
failing      A          87.000000      2
passing      A          87.166667      6
             B          85.250000      4

In [68]:
# I can even clean up my columns to make my calculations clearer.

grade_groups.columns = ['avg_reading_grade', 'count_of_students']
grade_groups

avg_reading_grade  count_of_students
passing_math classroom                                      
failing      A                  87.000000                  2
passing      A                  87.166667                  6
             B                  85.250000                  4

**Takeaways:**

We can interpret this output as there being 2 students failing math in classroom A with an average reading grade of 87, 6 students passing math in classroom A with an average reading grade of 87.16, and 4 students passing math in classroom B with an average reading grade of 85.25.

## Merging and Joining

Pandas provides several ways to combine dataframes together. We will look at two of them below:

### `pd.concat`

This function takes in a list or dictionary of Series or DataFrame objects and joins them along a particular axis, row-wise axis=0 or column-wise axis=1.

```python
# For example, concat with a list of two DataFrames
pd.concat([df1, df2], axis=0)
```

- When your list contains at least one DataFrame, a DataFrame is returned.


- When concatenating only Series objects row-wise, axis=0, a Series is returned.


- When concatenating Series or DataFrames column-wise, axis=1, a DataFrame is returned.

```python
# Default is set to row-wise concatenation using an outer join.
pd.concat(objs, axis=0, join='outer')
```

When concatenating dataframes vertically, we basically are just adding more rows to an existing dataframe. In this case, the dataframes we are putting together should have the same column names[^1].

In [69]:
"con" + "cat" + "e" + "nation"

'concatenation'

In [70]:
df1 = pd.DataFrame({'a': [1, 2, 3]})
df2 = pd.DataFrame({'a': [4, 5, 6]})

df1

a
0  1
1  2
2  3

In [71]:
df2

a
0  4
1  5
2  6

In [72]:
pd.concat([df1, df2])

a
0  1
1  2
2  3
0  4
1  5
2  6

**Note** that the indices are preserved on the resulting dataframe; we could set the `ignore_index` parameter to `True` if we wanted these to be sequential.

In [73]:
concat_df1 = pd.concat([df1, df2], ignore_index=True)
concat_df1

a
0  1
1  2
2  3
3  4
4  5
5  6

[^1]:
    We can concatenate dataframes with different column names, but generally this is not the behavior we want, as pandas will fill in a lot of null values into the resulting dataframe. The exception to this is if the dataframes are aligned on their index (i.e. the labels for each row), then we can provide the `axis=1` keyword argument to `pd.concat` to merge the dataframes horizontally.

In [74]:
concat_df2 = pd.DataFrame({'b': [1, 2, 3, 4, 5, 6]})
concat_df2

b
0  1
1  2
2  3
3  4
4  5
5  6

In [75]:
# Axis=1 means concat horizontally
pd.concat([concat_df1, concat_df2], axis=1)

a  b
0  1  1
1  2  2
2  3  3
3  4  4
4  5  5
5  6  6

### `.merge`

This method is similar to a SQL join. Here's a [cool read](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html#compare-with-sql-join) making a comparison between the two, if you're interested.

```python
# df.merge default settings for commonly used parameters.

left_df.merge(right_df, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, indicator=False)
```

How does changing the default argument of the `how` parameter change my resulting DataFrame?

`how` == Type of merge to be performed.

`how=left`: use only keys from left frame, similar to a SQL left outer join; preserve key order.

`how=right`: use only keys from right frame, similar to a SQL right outer join; preserve key order.

`how=outer`: use union of keys from both frames, similar to a SQL full outer join; sort keys lexicographically.

`how=inner`: use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys.

In [76]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [77]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

The `.merge` method will allow us to specify `left_on` and `right_on` to indicate the columns that are the keys used to merge the dataframes together. 

- In addition, the `how` keyword argument is used to define what type of JOIN we want to do; as we saw above, `inner` is the default setting. 

- For demonstration purposes, I'm also going to set the `indicator` parameter to `True`, which will create a column indicating whether the merge key appears in the `left_only`, `right_only` or `both` DataFrames.

In [78]:
# Perform an outer join specifying the left and right DataFrame keys.

users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

Notice that we have duplicate column names in the resulting dataframe. By default, pandas will add a suffix of `_x` to any columns in the left dataframe that are duplicated, and `_y` to any columns in the right dataframe that are duplicated. I can clean up my columns if I want to; one way would be to use method chaining, which it demonstrated below:

In [79]:
# Inner join tolerates no nulls.
# Inner join only shows user who have roles, roles that have users
users.merge(roles, how="inner", left_on="role_id", right_on="id")

id_x name_x  role_id  id_y    name_y
0     1    bob      1.0     1     admin
1     2    joe      2.0     2    author
2     3  sally      3.0     3  reviewer
3     4   adam      3.0     3  reviewer

In [80]:
(users.merge(roles, 
            left_on='role_id', 
            right_on='id', 
            how='outer')
    .drop(columns='role_id')
    .rename(columns={'id_x': 'id', 
                     'name_x': 'employee',
                     'id_y': 'role_id',
                     'name_y': 'role'}
            )
)

id employee  role_id       role
0  1.0      bob      1.0      admin
1  2.0      joe      2.0     author
2  3.0    sally      3.0   reviewer
3  4.0     adam      3.0   reviewer
4  5.0     jane      NaN        NaN
5  6.0     mike      NaN        NaN
6  NaN      NaN      4.0  commenter

## Exercises II

1. Copy the `users` and `roles` DataFrames from the examples above. 

2. What is the result of using a `right` join on the DataFrames? 

3. What is the result of using an `outer` join on the DataFrames?
     
4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

5. Load the `mpg` dataset from PyDataset. 

6. Output and read the documentation for the `mpg` dataset.

7. How many rows and columns are in the dataset?

8. Check out your column names and perform any cleanup you may want on them.

9. Display the summary statistics for the dataset.

10. How many different manufacturers are there?

11. How many different models are there?

12. Create a column named `mileage_difference` like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

13. Create a column named `average_mileage` like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

14. Create a new column on the `mpg` dataset named `is_automatic` that holds boolean values denoting whether the car has an automatic transmission.

15. Using the `mpg` dataset, find out which which manufacturer has the best miles per gallon on average?

16. Do automatic or manual cars have better miles per gallon?

In [81]:
# Create the users DataFrame. #1

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [82]:
# Create the roles DataFrame #1

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

In [83]:
# left_df.merge(right_df, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, indicator=False)

In [84]:
users.merge(roles, how ='right', left_on="role_id", right_on="id")

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0     1      admin
1   2.0    joe      2.0     2     author
2   3.0  sally      3.0     3   reviewer
3   4.0   adam      3.0     3   reviewer
4   NaN    NaN      NaN     4  commenter

In [85]:
users.merge(roles, how ='right') #it only merged the left primary key with the right 

id       name  role_id
0   1      admin      NaN
1   2     author      NaN
2   3   reviewer      NaN
3   4  commenter      NaN

In [86]:
users.merge(roles, how ='outer', left_on="role_id", right_on="id")

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0   1.0      admin
1   2.0    joe      2.0   2.0     author
2   3.0  sally      3.0   3.0   reviewer
3   4.0   adam      3.0   3.0   reviewer
4   5.0   jane      NaN   NaN        NaN
5   6.0   mike      NaN   NaN        NaN
6   NaN    NaN      NaN   4.0  commenter

In [87]:
users.merge(roles, how ='outer') #it only merged the primary key of hte secondary table

id       name  role_id
0   1        bob      1.0
1   2        joe      2.0
2   3      sally      3.0
3   4       adam      3.0
4   5       jane      NaN
5   6       mike      NaN
6   1      admin      NaN
7   2     author      NaN
8   3   reviewer      NaN
9   4  commenter      NaN

In [88]:
from pydataset import data

In [89]:
data('mpg', show_doc = True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [90]:
mpg = data('mpg') #5

In [91]:
mpg.head() #6

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [92]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

In [93]:
mpg.shape #7

(234, 11)

In [94]:
mpg.describe() #9

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

In [95]:
mpg.manufacturer.unique()

array(['audi', 'chevrolet', 'dodge', 'ford', 'honda', 'hyundai', 'jeep',
       'land rover', 'lincoln', 'mercury', 'nissan', 'pontiac', 'subaru',
       'toyota', 'volkswagen'], dtype=object)

In [96]:
len(mpg.manufacturer.unique())

15

In [97]:
mpg.model.unique()

array(['a4', 'a4 quattro', 'a6 quattro', 'c1500 suburban 2wd', 'corvette',
       'k1500 tahoe 4wd', 'malibu', 'caravan 2wd', 'dakota pickup 4wd',
       'durango 4wd', 'ram 1500 pickup 4wd', 'expedition 2wd',
       'explorer 4wd', 'f150 pickup 4wd', 'mustang', 'civic', 'sonata',
       'tiburon', 'grand cherokee 4wd', 'range rover', 'navigator 2wd',
       'mountaineer 4wd', 'altima', 'maxima', 'pathfinder 4wd',
       'grand prix', 'forester awd', 'impreza awd', '4runner 4wd',
       'camry', 'camry solara', 'corolla', 'land cruiser wagon 4wd',
       'toyota tacoma 4wd', 'gti', 'jetta', 'new beetle', 'passat'],
      dtype=object)

In [98]:
len(mpg.model.unique())

38

In [99]:
mpg.model.nunique()

38

In [100]:
mpg['milage_difference'] = (mpg.hwy - mpg.cty)

In [101]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   milage_difference  
1                 11  
2                  8  
3                 11  
4                  9  
5                 10

In [102]:
mpg['average_milage'] = (mpg.hwy + mpg.cty)/2

In [103]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   milage_difference  average_milage  
1                 11            23.5  
2                  8            25.0  
3                 11            25.5  
4                  9            25.5  
5                 10            21.0

In [104]:
mpg.dtypes

manufacturer          object
model                 object
displ                float64
year                   int64
cyl                    int64
trans                 object
drv                   object
cty                    int64
hwy                    int64
fl                    object
class                 object
milage_difference      int64
average_milage       float64
dtype: object

In [105]:
mpg[mpg.trans.str.startswith('auto')]

manufacturer                   model  displ  year  cyl     trans drv  cty  \
1           audi                      a4    1.8  1999    4  auto(l5)   f   18   
4           audi                      a4    2.0  2008    4  auto(av)   f   21   
5           audi                      a4    2.8  1999    6  auto(l5)   f   16   
7           audi                      a4    3.1  2008    6  auto(av)   f   18   
9           audi              a4 quattro    1.8  1999    4  auto(l5)   4   16   
11          audi              a4 quattro    2.0  2008    4  auto(s6)   4   19   
12          audi              a4 quattro    2.8  1999    6  auto(l5)   4   15   
14          audi              a4 quattro    3.1  2008    6  auto(s6)   4   17   
16          audi              a6 quattro    2.8  1999    6  auto(l5)   4   15   
17          audi              a6 quattro    3.1  2008    6  auto(s6)   4   17   
18          audi              a6 quattro    4.2  2008    8  auto(s6)   4   16   
19     chevrolet      c1500 suburban 2wd    5.3  2008    8  auto(l4)   r   14   
20     chevrolet      c1500 suburban 2wd    5.3  2008    8  auto(l4)   r   11   
21     chevrolet      c1500 suburban 2wd    5.3  2008    8  auto(l4)   r   14   
22     chevrolet      c1500 suburban 2wd    5.7  1999    8  auto(l4)   r   13   
23     chevrolet      c1500 suburban 2wd    6.0  2008    8  auto(l4)   r   12   
25     chevrolet                corvette    5.7  1999    8  auto(l4)   r   15   
27     chevrolet                corvette    6.2  2008    8  auto(s6)   r   15   
29     chevrolet         k1500 tahoe 4wd    5.3  2008    8  auto(l4)   4   14   
30     chevrolet         k1500 tahoe 4wd    5.3  2008    8  auto(l4)   4   11   
31     chevrolet         k1500 tahoe 4wd    5.7  1999    8  auto(l4)   4   11   
32     chevrolet         k1500 tahoe 4wd    6.5  1999    8  auto(l4)   4   14   
33     chevrolet                  malibu    2.4  1999    4  auto(l4)   f   19   
34     chevrolet                  malibu    2.4  2008    4  auto(l4)   f   22   
35     chevrolet                  malibu    3.1  1999    6  auto(l4)   f   18   
36     chevrolet                  malibu    3.5  2008    6  auto(l4)   f   18   
37     chevrolet                  malibu    3.6  2008    6  auto(s6)   f   17   
38         dodge             caravan 2wd    2.4  1999    4  auto(l3)   f   18   
39         dodge             caravan 2wd    3.0  1999    6  auto(l4)   f   17   
40         dodge             caravan 2wd    3.3  1999    6  auto(l4)   f   16   
41         dodge             caravan 2wd    3.3  1999    6  auto(l4)   f   16   
42         dodge             caravan 2wd    3.3  2008    6  auto(l4)   f   17   
43         dodge             caravan 2wd    3.3  2008    6  auto(l4)   f   17   
44         dodge             caravan 2wd    3.3  2008    6  auto(l4)   f   11   
45         dodge             caravan 2wd    3.8  1999    6  auto(l4)   f   15   
46         dodge             caravan 2wd    3.8  1999    6  auto(l4)   f   15   
47         dodge             caravan 2wd    3.8  2008    6  auto(l6)   f   16   
48         dodge             caravan 2wd    4.0  2008    6  auto(l6)   f   16   
50         dodge       dakota pickup 4wd    3.7  2008    6  auto(l4)   4   14   
51         dodge       dakota pickup 4wd    3.9  1999    6  auto(l4)   4   13   
53         dodge       dakota pickup 4wd    4.7  2008    8  auto(l5)   4   14   
54         dodge       dakota pickup 4wd    4.7  2008    8  auto(l5)   4   14   
55         dodge       dakota pickup 4wd    4.7  2008    8  auto(l5)   4    9   
57         dodge       dakota pickup 4wd    5.2  1999    8  auto(l4)   4   11   
58         dodge             durango 4wd    3.9  1999    6  auto(l4)   4   13   
59         dodge             durango 4wd    4.7  2008    8  auto(l5)   4   13   
60         dodge             durango 4wd    4.7  2008    8  auto(l5)   4    9   
61         dodge             durango 4wd    4.7  2008    8  auto(l5)   4   13   
62         dodge             durang

In [106]:

#You can chain startswith and endswith masks or use contains - ^ is for match start of string, .* is for any string and $ for end
# % did not work as wildcard as I used it

In [107]:
mpg['is_automatic'] = mpg.trans.str.startswith('auto')

In [108]:
#mpg['is_automatic'] = mpg.trans.str.contains('auto')

In [109]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   milage_difference  average_milage  is_automatic  
1                 11            23.5          True  
2                  8            25.0         False  
3                 11            25.5         False  
4                  9            25.5          True  
5                 10            21.0          True

In [110]:
mpg.groupby('manufacturer')

In [111]:
mpg.groupby('manufacturer')[['average_milage']].mean().sort_values('average_milage',ascending = False)

average_milage
manufacturer                
honda              28.500000
volkswagen         25.074074
hyundai            22.750000
subaru             22.428571
audi               22.027778
toyota             21.720588
pontiac            21.700000
nissan             21.346154
chevrolet          18.447368
ford               16.680000
mercury            15.625000
jeep               15.562500
dodge              15.540541
lincoln            14.166667
land rover         14.000000

In [112]:
#class
mpg.groupby('manufacturer').average_milage.mean().nlargest(n=1)

manufacturer
honda    28.5
Name: average_milage, dtype: float64

In [113]:
#16 Do automatic or manual cars have better miles per gallon?

In [114]:
mpg.groupby('is_automatic')[['average_milage']].mean()

average_milage
is_automatic                
False              22.227273
True               19.130573

## Making it Pretty 
    -method 1, change index
    -method 2, use np.where, update table itself

In [115]:
n_mpg = mpg.groupby('is_automatic')[['average_milage']].mean()

In [116]:
n_mpg.index

Index([False, True], dtype='object', name='is_automatic')

In [117]:
n_mpg.index = ['manual', 'auto']

In [118]:
n_mpg.index

Index(['manual', 'auto'], dtype='object')

In [119]:
n_mpg.head()

average_milage
manual       22.227273
auto         19.130573

In [120]:
mpg['is_automatic2'] = np.where(mpg.trans.str.startswith('auto'), 'Auto', 'Manual')

In [121]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   milage_difference  average_milage  is_automatic is_automatic2  
1                 11            23.5          True          Auto  
2                  8            25.0         False        Manual  
3                 11            25.5         False        Manual  
4                  9            25.5          True          Auto  
5                 10            21.0          True          Auto

In [122]:
mpg.groupby('is_automatic2')[['average_milage']].mean()

average_milage
is_automatic2                
Auto                19.130573
Manual              22.227273

## Reshaping

We will talk about reshaping operations in more detail when we discuss tidy data, but for now we will focus on a couple of common operations that can be used to summarize our data by different subgroups.

### `pd.crosstab`

For an example of `.crosstab`, we will count the number of students passing math in each classroom.

In [123]:
# We will use our student grades DataFrame, df.

df

name  math  english  reading classroom passing_math
0     Sally    62       85       80         A      failing
1      Jane    88       79       67         B      passing
2     Suzie    94       74       95         A      passing
3     Billy    98       96       88         B      passing
4       Ada    77       92       98         A      passing
5      John    79       76       93         B      passing
6    Thomas    82       64       81         A      passing
7     Marie    93       63       90         A      passing
8    Albert    92       62       87         A      passing
9   Richard    69       80       94         A      failing
10    Isaac    92       99       93         B      passing
11     Alan    92       62       72         A      passing

We'll use the `pd.crosstab` function to count the number of occurances of each subgroup (i.e. each unique combination of classroom and whether or not the student is passing math):

In [124]:
pd.crosstab(df.passing_math, df.classroom)

classroom     A  B
passing_math      
failing       2  0
passing       6  4

We can also view subtotals with the `margins` set to `True`.

In [125]:
pd.crosstab(df.passing_math, df.classroom, margins=True)

classroom     A  B  All
passing_math           
failing       2  0    2
passing       6  4   10
All           8  4   12

The `.crosstab` function will let us view the numbers as percentages of the total as well by setting `normalize` to `True`.

In [126]:
pd.crosstab(df.passing_math, df.classroom, normalize=True).round(3)

classroom         A      B
passing_math              
failing       0.167  0.000
passing       0.500  0.333

In [127]:
pd.crosstab(df.passing_math, df.classroom, normalize=True ).round(3)

classroom         A      B
passing_math              
failing       0.167  0.000
passing       0.500  0.333

In [128]:
pd.crosstab(df.passing_math, df.classroom, normalize= 'index' ).round(3)

classroom       A    B
passing_math          
failing       1.0  0.0
passing       0.6  0.4

In [129]:
pd.crosstab(df.passing_math, df.classroom, normalize= 'columns' ).round(3)

classroom        A    B
passing_math           
failing       0.25  0.0
passing       0.75  1.0

### `.pivot_table`

Here we use the `.pivot_table` method to create our summary. This method produces output similar to an excel pivot table. We must supply 3 things here:

- which values will make up the rows (the `index`)
- which values will make up the columns
- the values we are aggregating
- an aggregation method (`aggfunc`); if we can omit this, and `mean` will be used by default

For an example using the `pivot_table` method, we'll calculate the average math grade for the combination of `classroom` and `passing_math` status.

In [130]:
df.pivot_table(index='classroom', columns='passing_math', values='math')

passing_math  failing    passing
classroom                       
A                65.5  88.333333
B                 NaN  89.250000

Here we'll create a dataframe that represents various orders at a restaurant.

In [131]:
n = 40

orders = pd.DataFrame({
    'drink': np.random.choice(['Tea', 'Water', 'Water'], n),
    'meal': np.random.choice(['Curry', 'Yakisoba Noodle', 'Pad Thai'], n),
})

orders.sample(10)

drink             meal
0     Tea         Pad Thai
38  Water            Curry
11    Tea            Curry
4     Tea            Curry
16  Water            Curry
2     Tea  Yakisoba Noodle
22  Water  Yakisoba Noodle
5   Water         Pad Thai
19    Tea         Pad Thai
15  Water  Yakisoba Noodle

#### `.map`

The `.map` method lets us use a dictionary to calculate the total price for an order; then I can save my calculations to a new column named `bill`. Let's do this step-by-step.

In [132]:
# Create a dictionary of prices for drinks and meals.

prices = {
    'Yakisoba Noodle': 9,
    'Curry': 11,
    'Pad Thai': 10,
    'Tea': 2,
    'Water': 0,
}

In [133]:
orders.meal

0            Pad Thai
1            Pad Thai
2     Yakisoba Noodle
3            Pad Thai
4               Curry
5            Pad Thai
6            Pad Thai
7     Yakisoba Noodle
8     Yakisoba Noodle
9            Pad Thai
10           Pad Thai
11              Curry
12    Yakisoba Noodle
13    Yakisoba Noodle
14              Curry
15    Yakisoba Noodle
16              Curry
17    Yakisoba Noodle
18           Pad Thai
19           Pad Thai
20              Curry
21    Yakisoba Noodle
22    Yakisoba Noodle
23    Yakisoba Noodle
24              Curry
25    Yakisoba Noodle
26    Yakisoba Noodle
27              Curry
28           Pad Thai
29    Yakisoba Noodle
30           Pad Thai
31              Curry
32    Yakisoba Noodle
33              Curry
34           Pad Thai
35              Curry
36              Curry
37           Pad Thai
38              Curry
39           Pad Thai
Name: meal, dtype: object

In [134]:
orders.meal.map(prices)

0     10
1     10
2      9
3     10
4     11
5     10
6     10
7      9
8      9
9     10
10    10
11    11
12     9
13     9
14    11
15     9
16    11
17     9
18    10
19    10
20    11
21     9
22     9
23     9
24    11
25     9
26     9
27    11
28    10
29     9
30    10
31    11
32     9
33    11
34    10
35    11
36    11
37    10
38    11
39    10
Name: meal, dtype: int64

In [135]:
"""
Match the values in the 'drink' and 'meal' columns with the values in the 'prices' dictionary 
and perform the specified calculation. Save this calculation to a new column named 'bill'.
"""

orders['bill'] = orders.drink.map(prices) + orders.meal.map(prices)

orders.sample(10)

drink             meal  bill
6   Water         Pad Thai    10
31  Water            Curry    11
7   Water  Yakisoba Noodle     9
32  Water  Yakisoba Noodle     9
0     Tea         Pad Thai    12
22  Water  Yakisoba Noodle     9
38  Water            Curry    11
15  Water  Yakisoba Noodle     9
21    Tea  Yakisoba Noodle    11
29    Tea  Yakisoba Noodle    11

Let's take a look at how many orders have each combination of meal and drink:

In [136]:
pd.crosstab(orders.drink, orders.meal)

meal   Curry  Pad Thai  Yakisoba Noodle
drink                                  
Tea        3         4                5
Water      9        10                9

In [137]:
pd.crosstab(orders.drink, orders.meal, normalize=True, margins=True)

meal   Curry  Pad Thai  Yakisoba Noodle  All
drink                                       
Tea    0.075      0.10            0.125  0.3
Water  0.225      0.25            0.225  0.7
All    0.300      0.35            0.350  1.0

In [138]:
pd.crosstab(orders.drink, orders.meal, normalize=True, margins=False)

meal   Curry  Pad Thai  Yakisoba Noodle
drink                                  
Tea    0.075      0.10            0.125
Water  0.225      0.25            0.225

And let's find out the average bill amount for each combination: 

In [139]:
orders.pivot_table(index='drink', columns='meal', values='bill')

meal   Curry  Pad Thai  Yakisoba Noodle
drink                                  
Tea       13        12               11
Water     11        10                9

It's interesting to note that we could find the same information with a multi-level group by:

In [140]:
orders.groupby(['drink', 'meal']).bill.mean()

drink  meal           
Tea    Curry              13.0
       Pad Thai           12.0
       Yakisoba Noodle    11.0
Water  Curry              11.0
       Pad Thai           10.0
       Yakisoba Noodle     9.0
Name: bill, dtype: float64

The choice between group by and a pivot table here is mostly asthetic, and you should use whichever makes more sense to you with the problem at hand. 

### Transposing

In [141]:
df.T

0        1        2        3        4        5        6   \
name            Sally     Jane    Suzie    Billy      Ada     John   Thomas   
math               62       88       94       98       77       79       82   
english            85       79       74       96       92       76       64   
reading            80       67       95       88       98       93       81   
classroom           A        B        A        B        A        B        A   
passing_math  failing  passing  passing  passing  passing  passing  passing   

                   7        8        9        10       11  
name            Marie   Albert  Richard    Isaac     Alan  
math               93       92       69       92       92  
english            63       62       80       99       62  
reading            90       87       94       93       72  
classroom           A        A        A        B        A  
passing_math  passing  passing  failing  passing  passing

In [142]:
df.describe().T

count       mean        std   min    25%   50%    75%   max
math      12.0  84.833333  11.134168  62.0  78.50  90.0  92.25  98.0
english   12.0  77.666667  13.371158  62.0  63.75  77.5  86.75  99.0
reading   12.0  86.500000   9.643651  67.0  80.75  89.0  93.25  98.0

## Exercises III

1. Use your `get_db_url` function to help you explore the data from the `chipotle` database. 

2. What is the total price for each order?

3. What are the most popular 3 items?

4. Which item has produced the most revenue?

5. Join the `employees` and `titles` DataFrames together.

6. For each title, find the hire date of the employee that was hired most recently with that title.

7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [8]:
import env

In [9]:
pd.read_sql('show tables', env.get_db_url('chipotle')) #from class

Tables_in_chipotle
0             orders

In [10]:
chp = pd.read_sql('SELECT * FROM orders', env.get_db_url('chipotle') )

In [11]:
chp.head(3)

id  order_id  quantity                     item_name choice_description  \
0   1         1         1  Chips and Fresh Tomato Salsa                nan   
1   2         1         1                          Izze       [Clementine]   
2   3         1         1              Nantucket Nectar            [Apple]   

  item_price  
0     $2.39   
1     $3.39   
2     $3.39

In [12]:
chp.shape

(4622, 6)

In [13]:
chp.dtypes

id                     int64
order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

In [14]:
chp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  4622 non-null   int64 
 1   order_id            4622 non-null   int64 
 2   quantity            4622 non-null   int64 
 3   item_name           4622 non-null   object
 4   choice_description  4622 non-null   object
 5   item_price          4622 non-null   object
dtypes: int64(3), object(3)
memory usage: 216.8+ KB


In [15]:
chp.item_price = chp.item_price.str.replace('$', '').str.strip().astype(dtype = 'float')

/var/folders/lh/dkl_j44s5s192k6tt8pqtrdh0000gn/T/ipykernel_1774/433354939.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  chp.item_price = chp.item_price.str.replace('$', '').str.strip().astype(dtype = 'float')


In [16]:
chp.head(3)

id  order_id  quantity                     item_name choice_description  \
0   1         1         1  Chips and Fresh Tomato Salsa                nan   
1   2         1         1                          Izze       [Clementine]   
2   3         1         1              Nantucket Nectar            [Apple]   

   item_price  
0        2.39  
1        3.39  
2        3.39

In [17]:
chp.dtypes

id                      int64
order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

In [18]:
chp['total_per_item'] = (chp.item_price * chp.quantity)

In [19]:
chp.head(3)

id  order_id  quantity                     item_name choice_description  \
0   1         1         1  Chips and Fresh Tomato Salsa                nan   
1   2         1         1                          Izze       [Clementine]   
2   3         1         1              Nantucket Nectar            [Apple]   

   item_price  total_per_item  
0        2.39            2.39  
1        3.39            3.39  
2        3.39            3.39

In [20]:
chp.groupby('order_id').total_per_item.sum() #2 works
chp.groupby('order_id')[['total_per_item']].sum().head(3) #2 prettier

total_per_item
order_id                
1                  11.56
2                  33.96
3                  12.67

In [21]:
chp.groupby('item_name').quantity.sum().nlargest(3) #3 works

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

In [22]:
chp.groupby('item_name')[['quantity']].sum().sort_values('quantity', ascending = False).head(3) #3 prettier

quantity
item_name                    
Chicken Bowl              761
Chicken Burrito           591
Chips and Guacamole       506

In [23]:
total_sold= chp.groupby('item_name')[['quantity']].sum().sort_values('quantity', ascending = False)

In [24]:
total_sold_crosstab = pd.crosstab(chp.item_name, chp.quantity.sum()).sort_values(chp.quantity.sum(), ascending = False)
#use for catagorical, so not appropriate here, use groupby for this one

In [25]:
confusion = total_sold.merge(total_sold_crosstab, how = 'left', on = 'item_name').head(3)

In [26]:
chp.quantity.sum() #as you can see, group by is best for calculations, crosstab is best for catagorical

4972

In [27]:
confusion.quantity.sum()

1858

In [28]:
confusion[4972].sum()

1758

In [29]:
chp.head(5)

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                  choice_description  item_price  \
0                                                nan        2.39   
1                                       [Clementine]        3.39   
2                                            [Apple]        3.39   
3                                                nan        2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98   

   total_per_item  
0            2.39  
1            3.39  
2            3.39  
3            2.39  
4           33.96

In [30]:
chp[chp.item_name == 'Bowl']

id  order_id  quantity item_name  \
445  446       193         3      Bowl   
673  674       279         1      Bowl   

                                    choice_description  item_price  \
445  [Braised Carnitas, Pinto Beans, [Sour Cream, C...        22.2   
673  [Adobo-Marinated and Grilled Steak, [Sour Crea...         7.4   

     total_per_item  
445            66.6  
673             7.4

In [31]:
avg_cost_per_item = chp.groupby('item_name')[['item_price']].mean()

avg_cost_per_item.head(3)

item_price
item_name                    
6 Pack Soft Drink    6.610185
Barbacoa Bowl       10.187273
Barbacoa Burrito     9.832418

In [32]:
rev = total_sold.merge(avg_cost_per_item, how = 'left', on = 'item_name')

In [33]:
rev.head(3)

quantity  item_price
item_name                                
Chicken Bowl              761   10.113953
Chicken Burrito           591   10.082857
Chips and Guacamole       506    4.595073

In [34]:
rev['total_rev'] = (rev.quantity * rev.item_price)
rev.head(3)

quantity  item_price    total_rev
item_name                                             
Chicken Bowl              761   10.113953  7696.718361
Chicken Burrito           591   10.082857  5958.968571
Chips and Guacamole       506    4.595073  2325.106973

In [35]:
rev.sort_values('total_rev', ascending = False).head(3)

quantity  item_price    total_rev
item_name                                         
Chicken Bowl          761   10.113953  7696.718361
Chicken Burrito       591   10.082857  5958.968571
Steak Burrito         386   10.465842  4039.815163

In [36]:
#4 class

In [47]:
chp.groupby('item_name').item_price.sum().nlargest(n=1) #4 class interpretation

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

In [ ]:
#4 class interpretation did mention there are multiple logical ways to do this that will give separate answers
#I will be keeping my answer

In [38]:
emp = pd.read_sql('SELECT * FROM employees', env.get_db_url('employees') )

In [39]:
titles = pd.read_sql('SELECT * FROM titles', env.get_db_url('employees') )

In [40]:
emp_titles = emp.merge(titles, how ='left', on='emp_no')
emp_titles.head()

emp_no  birth_date first_name last_name gender   hire_date  \
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   

             title   from_date     to_date  
0  Senior Engineer  1986-06-26  9999-01-01  
1            Staff  1996-08-03  9999-01-01  
2  Senior Engineer  1995-12-03  9999-01-01  
3         Engineer  1986-12-01  1995-12-01  
4  Senior Engineer  1995-12-01  9999-01-01

In [41]:
emp_titles.groupby('title')[['hire_date']].max()

hire_date
title                         
Assistant Engineer  1999-12-24
Engineer            2000-01-28
Manager             1992-02-05
Senior Engineer     2000-01-01
Senior Staff        2000-01-13
Staff               2000-01-12
Technique Leader    1999-12-31

In [48]:
#class
emp_titles.groupby('title').hire_date.max().reset_index()

title   hire_date
0  Assistant Engineer  1999-12-24
1            Engineer  2000-01-28
2             Manager  1992-02-05
3     Senior Engineer  2000-01-01
4        Senior Staff  2000-01-13
5               Staff  2000-01-12
6    Technique Leader  1999-12-31

In [42]:
#side project

In [43]:
side = emp_titles.groupby('title')[['hire_date']].max()
side

hire_date
title                         
Assistant Engineer  1999-12-24
Engineer            2000-01-28
Manager             1992-02-05
Senior Engineer     2000-01-01
Senior Staff        2000-01-13
Staff               2000-01-12
Technique Leader    1999-12-31

In [44]:
side_quest_complete = side.merge(emp_titles, how = 'left', on = ['hire_date', 'title'])
side_quest_complete

hire_date               title  emp_no  birth_date   first_name last_name  \
0  1999-12-24  Assistant Engineer  243702  1961-10-04     Munehiro      Luke   
1  2000-01-28            Engineer  463807  1964-06-12       Bikash    Covnot   
2  1992-02-05             Manager  110420  1963-07-27        Oscar  Ghazalie   
3  2000-01-01     Senior Engineer  108201  1955-04-14  Mariangiola   Boreale   
4  2000-01-13        Senior Staff  222965  1959-08-07      Volkmar     Perko   
5  2000-01-12               Staff   47291  1960-09-09          Ulf    Flexer   
6  1999-12-31    Technique Leader  294732  1960-11-07       Karlis    Orsini   

  gender   from_date     to_date  
0      F  2000-01-07  9999-01-01  
1      M  2000-02-01  2000-05-19  
2      M  1996-08-30  9999-01-01  
3      M  2000-01-01  9999-01-01  
4      F  2000-01-29  9999-01-01  
5      M  2000-01-26  9999-01-01  
6      M  2000-01-28  2001-07-04

In [3]:
query = '''
SELECT
    d.dept_name,
    t.title
FROM departments d
JOIN dept_emp USING (dept_no)
JOIN titles t USING (emp_no)
'''

titles_again = pd.read_sql(query, env.get_db_url('employees'))
titles_again

dept_name         title
0       Customer Service         Staff
1       Customer Service  Senior Staff
2       Customer Service         Staff
3       Customer Service  Senior Staff
4       Customer Service         Staff
...                  ...           ...
489898             Sales  Senior Staff
489899             Sales         Staff
489900             Sales         Staff
489901             Sales  Senior Staff
489902             Sales         Staff

[489903 rows x 2 columns]

In [8]:
#class
query = '''
SELECT
    d.dept_name,
    t.title
FROM departments d
JOIN dept_emp de USING (dept_no)
JOIN titles t USING (emp_no)

where de.to_date > now() and t.to_date > now()
'''

titles_again2 = pd.read_sql(query, env.get_db_url('employees'))
titles_again2

dept_name         title
0       Customer Service  Senior Staff
1       Customer Service  Senior Staff
2       Customer Service  Senior Staff
3       Customer Service  Senior Staff
4       Customer Service         Staff
...                  ...           ...
240119             Sales  Senior Staff
240120             Sales  Senior Staff
240121             Sales  Senior Staff
240122             Sales  Senior Staff
240123             Sales         Staff

[240124 rows x 2 columns]

In [11]:
#class
query = '''
SELECT
    d.dept_name,
    t.title
FROM departments d
JOIN dept_emp de USING (dept_no)
JOIN titles t USING (emp_no)

where de.to_date < '9999-01-01' and t.to_date < '9999-01-01';
'''

titles_again2 = pd.read_sql(query, env.get_db_url('employees'))
titles_again2

dept_name            title
0      Customer Service            Staff
1      Customer Service         Engineer
2      Customer Service  Senior Engineer
3      Customer Service            Staff
4      Customer Service     Senior Staff
...                 ...              ...
93447             Sales            Staff
93448             Sales     Senior Staff
93449             Sales     Senior Staff
93450             Sales     Senior Staff
93451             Sales            Staff

[93452 rows x 2 columns]

In [ ]:
#idk why the two above tables are different

In [12]:
pd.crosstab(titles_again2.dept_name, titles_again.title)

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                    51       466        0              394   
Development                        205      1599        4             1376   
Finance                             94       671        0              566   
Human Resources                    307      2353        0             1945   
Marketing                          332      2519        0             2249   
Production                        1270      9495        0             8098   
Quality Management                 393      2603        0             2238   
Research                           379      2757        2             2320   
Sales                              935      6799        0             5819   

title               Senior Staff  Staff  Technique Leader  
dept_name                                                  
Customer Service            2499   2894                53  
Development                 9761  11255               214  
Finance                     1675   2010                98  
Human Resources               53     61               303  
Marketing                     66     64               320  
Production                   199    223              1280  
Quality Management            58     70               328  
Research                      62     78               356  
Sales                        151    160               922

In [ ]:
#altn solution: groupby

#NOTE: in class he also imported emp_no in his original sql pull, 
#I did not so I would have to remedy that to use the following methods below

In [53]:
#titles_again2.groupby(['title', 'dept_name']).emp_no.count() 

In [54]:
#titles_again2.groupby(['title', 'dept_name'])[[emp_no]].count()

### Extra Pandas Exercises and Resources
- [https://www.w3resource.com/python-exercises/pandas/index.php](https://www.w3resource.com/python-exercises/pandas/index.php)
- [https://towardsdatascience.com/20-pandas-functions-that-will-boost-your-data-analysis-process-f5dfdb2f9e05](https://towardsdatascience.com/20-pandas-functions-that-will-boost-your-data-analysis-process-f5dfdb2f9e05)
- [https://github.com/guipsamora/pandas_exercises](https://github.com/guipsamora/pandas_exercises)
- [https://github.com/ajcr/100-pandas-puzzles](https://github.com/ajcr/100-pandas-puzzles)

### More Practice!

For even more practice with pandas, you can do the exercises from the SQL module, but instead of using SQL to do the aggregation, sorting, joining, etc, use pandas. That is, read the data from all of the tables into pandas dataframes and manipulate the dataframes.